libraries/core/src/main/java/com/paritytrading/philadelphia/FIXValue.java

```diff
@@ -211,48 +232,87 @@ public void setInt(long x) {
     /**
      * Get the value as a float.
      *
+     * <p><strong>Note.</strong> The value is a string representation of
+     * a decimal number. As converting an arbitrary decimal number into a
+     * floating point number requires arbitrary-precision arithmetic, this
+     * method only works with the subset of decimal numbers that can be
+     * converted into floating point numbers using floating-point
+     * arithmetic.</p>
+     *
+     * <p>If we represent a decimal number in the form
+     *
+     *     &plusmn;<i>s</i>&nbsp;&times;&nbsp;10<sup><i>e</i></sup>,
+     *
+     * where <i>s</i> is an integer significand and <i>e</i> is an integer
+     * exponent, this method works for decimal numbers having
+     *
+     *     0&nbsp;&le;&nbsp;<i>s</i>&nbsp;&le;&nbsp;2<sup>53</sup>&nbsp;-&nbsp;1
+     *
+     * and
+     *
+     *     -22&nbsp;&le;&nbsp;<i>e</i>&nbsp;&le;&nbsp;22.</p>
+     *
      * @return the value as a float
      * @throws FIXValueFormatException if the value is not a float
+     * @see <a href="https://www.exploringbinary.com/fast-path-decimal-to-floating-point-conversion/">Fast Path Decimal to Floating-Point Conversion</a>
      */
     public double asFloat() {
-        long x = 0;
+        boolean negative = false;
+
+        int i = offset;
 
-        double factor = 0.0;
+        if (bytes[i] == '-') {
+            negative = true;
 
-        long sign  = bytes[offset] == '-' ? -1 : +1;
-        int  start = sign < 0 ? offset + 1 : offset;
+            i++;
+        }
 
-        for (int i = start; i < offset + length; i++) {
-            byte b = bytes[i];
+        long significand = 0;
 
-            if (b < '0' || b > '9') {
-                if (factor == 0.0 && b == '.') {
-                    factor = 1.0;
-                    continue;
-                }
+        byte b = 0;
 
+        while (i < offset + length) {
+            b = bytes[i++];
+
+            if (b < '0' || b > '9')
+                break;
+
+            significand = 10 * significand + b - '0';
+        }
+
+        if (b != '.' && i < offset + length)
+            notFloat();
+
+        int exponent = 0;
+
+        while (i < offset + length) {
+            b = bytes[i++];
+
+            if (b < '0' || b > '9')
                 notFloat();
-            }
 
-            x = 10 * x + b - '0';
+            significand = 10 * significand + b - '0';
 
-            factor *= 10;
+            exponent++;
         }
 
-        return sign * (factor > 0.0 ? x / factor : x);
+        double x = exponent != 0 ? significand / POWERS_OF_TEN[exponent] : significand;
```

Comment:
Could you drop this conditional and just write below as `POWERS_OF_TEN[0] == 1`:
```
double x = significand / POWERS_OF_TEN[exponent];
```

Maybe something for the benchmark to decide.

Reply:
Good catch! 🔍  We certainly don't want to optimize for `exponent == 0` at the expense of `exponent != 0`.

Reply:
This is now fixed. 👍